plone/app/z3cform/widget.py

```diff
@@ -1,612 +1,940 @@
+        else:
+            dm = None
+        ret_value = '<div class="files">'
+        current_field_value = (
+            dm.query()
+            if ((dm is not None) and
+                self.field.interface.providedBy(self.context))
+            else None
+        )
+        if current_field_value and current_field_value != NO_VALUE:
+            if not isinstance(current_field_value, list):
+                current_field_value = [current_field_value]
+            current_field_set = set(current_field_value)
+            for item in current_field_set:
+                ret_value = ret_value + '<div class="existfileupload">'
+                dl_url = '%s/++widget++%s/@@downloadexisting/' % (
+                             self.request.getURL(),
+                             self.name) + item.filename
+                ret_value = ret_value + '<a href=' + dl_url + '>'
+                ret_value = ret_value + '<span class="filename">' + item.filename + '</span>'
+                size = self.formatSize(item.getSize())
+                ret_value = ret_value + '<span class="filesize"> ' + size + '</span>'
+                ret_value = ret_value + '</div>'
+        ret_value = ret_value + '</div>'
+        return ret_value
+
+    def formatSize(self, numBytes):
+        """
+        Format a human readable file size
+        """
+        if numBytes > 1000000000:
+            return str(int(round(numBytes / (1024 * 1024 * 1024)))) + ' GB'
+        if numBytes > 1000000:
+            return str(int(round(numBytes / (1024 * 1024)))) + ' MB'
+        return str(int(round(numBytes / 1024))) + ' KB'
+
+    def cleanup(self):
+        """
+        look through upload directory and remove old uploads
+        (older than 2 hrs)
+        """
+        now = time.time()
+        tmpdir = gettempdir()
+        for filename in os.listdir(tmpdir):
+            if fnmatch.fnmatch(filename, '*FileUpload'):
+                filepath = os.path.join(tmpdir, filename)
+                if (os.stat(filepath).st_mtime) < now - 2 * 60 * 60:
+                    os.unlink(filepath)
+
+
+@implementer(IFieldWidget)
+def FileUploadFieldWidget(field, request):
+    return FieldWidget(field, FileUploadWidget(request))
+
+
+class Upload(BrowserView):
+    """Upload a file via ++widget++widget_name/@@upload"""
+
+    implements(IPublishTraverse)
+
+    def __call__(self):
+
+        if hasattr(self.request, "REQUEST_METHOD"):
+            # TODO: we should check errors in the creation process, and
+            # broadcast those to the error template in JS
+            if self.request["REQUEST_METHOD"] == "POST":
+                if getattr(self.request, self.context.name, None) is not None:
+                    files = self.request[self.context.name]
+                    uploaded = self.upload(files)
+                    if uploaded:
+                        return json.dumps(uploaded)
+                return json.dumps("")
+
+    def upload(self, item):

```

Comment:
Yes. The file is saved to a temporary file. In the case of an AddForm there
isn't a zodb object to upload the file to. We could upload it to some
arbitrary object like the site root but I think cleanup of files left over
from a canceled add would be a problem. Using plone.app.drafts might be an
option but do you want that dependancy for a widget?

On Wed, May 27, 2015 at 3:52 PM, Nathan Van Gheem <notifications@github.com>
wrote:

> In plone/app/z3cform/widget.py
> <https://github.com/plone/plone.app.z3cform/pull/20#discussion_r31184744>:
>
> > +    implements(IPublishTraverse)
> > +
> > +    def __call__(self):
> > +
> > +        if hasattr(self.request, "REQUEST_METHOD"):
> > +            # TODO: we should check errors in the creation process, and
> > +            # broadcast those to the error template in JS
> > +            if self.request["REQUEST_METHOD"] == "POST":
> > +                if getattr(self.request, self.context.name, None) is not None:
> > +                    files = self.request[self.context.name]
> > +                    uploaded = self.upload(files)
> > +                    if uploaded:
> > +                        return json.dumps(uploaded)
> > +                return json.dumps("")
> > +
> > +    def upload(self, item):
>
> So you're not saving the file in the zodb? The file is saved as a
> temporary file?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/plone/plone.app.z3cform/pull/20/files#r31184744>.
>